modyn/supervisor/internal/supervisor.py

```diff
@@ -0,0 +1,368 @@
+    while True:
+        num_pipelines = len(pipeline_process_dict)
+        if num_pipelines > 0:
+            logger.info(f"[{os.getpid()}][pipeline_monitor] {num_pipelines} pipelines registered")
+
+        # join all the terminated procs
+        num_active_pipeline_processes = len(mp.active_children())
+        if num_active_pipeline_processes > 0:
+            logger.info(f"[{os.getpid()}][pipeline_monitor] {num_active_pipeline_processes} pipeline processes running")
+
+        for p_id, p_info in pipeline_process_dict.items():
+            p_info.update_status()
+            if p_info.pipeline_exited:
+                proc_alive = p_info.process_handler.is_alive()
+                logger.info(
+                    f"""[{os.getpid()}][pipeline_monitor] pipeline {p_id}
+                         exit code {p_info.pipeline_exit_code.name},
+                         proc alive {proc_alive}"""
+                )
+                if not proc_alive:
+                    # TODO(#317): unregister pipeline when process terminates
+                    pass
+            else:
+                logger.info(f"[{os.getpid()}][pipeline_monitor] pipeline {p_id} still running")
+
+        time.sleep(10)
+
+
+class Supervisor:
+    # pylint: disable=too-many-instance-attributes
+    # This is a core class and we require the attributes.
+
+    # TODO(#63): Get these from Selector
+    supported_strategies: list[str] = [
+        "NewDataStrategy",
+        "FreshnessSamplingStrategy",
+        "CoresetStrategy",
+    ]
+
+    supported_evaluation_result_writers: dict = {"json": JsonResultWriter, "tensorboard": TensorboardResultWriter}
+
+    def __init__(
+        self,
+        modyn_config: dict,
+    ) -> None:
+        # TODO(#325): validate modyn_config
+        self.modyn_config = modyn_config
+        self._pipeline_process_dict: dict[int, PipelineInfo] = {}
+        self._manager = Manager()
+        self._next_pipeline_lock = self._manager.Lock()
+        self.grpc = GRPCHandler(self.modyn_config)
+        self.pipeline_monitor_thread: Optional[threading.Thread] = None
+
+        # TODO(#317): redesign tensorboard. ignore it for now
+        # if "tensorboard" in self.modyn_config:
+        #     port = self.modyn_config["tensorboard"]["port"]
+        #     self._run_tensorboard(port)
+        #     logger.info(f"Starting up tensorboard on port {port}.")
+
+    def __getstate__(self) -> dict:
+        state = self.__dict__.copy()
+        if "_manager" in state:
+            del state["_manager"]
+        else:
+            logger.info("'_manager' not found in state")
+        return state
+
+    def init_metadata_db(self) -> None:
+        with MetadataDatabaseConnection(self.modyn_config) as database:
+            database.create_tables()
+
+    def init_cluster_connection(self) -> None:
+        logging.info("Setting up connections to cluster components.")
+        self.grpc.init_cluster_connection()
+        self.init_metadata_db()
```

Comment:
We might be able to init the metadata db in the constructor, since we only need to do that once

Reply:
done